File: Java/opentraceability/src/main/java/opentraceability/utility/attributes/OpenTraceabilityAttribute.java
package opentraceability.utility.attributes;

import opentraceability.models.events.EPCISVersion;

import javax.xml.namespace.QName;
import java.lang.annotation.*;

@Repeatable(OpenTraceabilityAttributes.class)
@Target(ElementType.FIELD)
@Retention(RetentionPolicy.RUNTIME)
public @interface OpenTraceabilityAttribute {
    String ns();

    String name();

    int sequenceOrder() default -1;

    EPCISVersion version() default EPCISVersion.Any;
}

File: Java/OpenTraceability/src/main/opentraceability/utility/attributes/opentraceabilityattribute.kt
package utility.attributes

import models.identifiers.*
import models.events.*
import models.events.EPCISVersion
import java.lang.reflect.Type
import javax.xml.namespace.QName

@Target(AnnotationTarget.PROPERTY)
@Retention(AnnotationRetention.RUNTIME)
annotation class OpenTraceabilityAttribute(
    val ns: String,
    val name: String,
    val sequenceOrder: Int = -1,
    val version: EPCISVersion = EPCISVersion.V1
)

fun getXNamespace(ns: String): String {
    return ns // Replace this with the appropriate logic to convert the namespace to XNamespace.
}

fun buildQualifiedName(ns: String, name: String): String {
    val namespace = getXNamespace(ns)
    return QName(namespace, name).toString()
}